worlds/sc2/rules.py

```diff
@@ -29,6 +35,121 @@ def is_item_placement(self, state):
+            else self.world.custom_mission_order.get_mission_count()
+
+    # Unit classes in world, these are set properly into the world after the item culling is done
+    # Therefore need to get from the world
+    def world_has_barracks_unit(self):
+        if self.world is None:
+            return False
+        else:
+            return self.world.has_barracks_unit
+
+    def world_has_factory_unit(self):
+        if self.world is None:
+            return False
+        else:
+            return self.world.has_factory_unit
+
+    def world_has_starport_unit(self):
+        if self.world is None:
+            return False
+        else:
+            return self.world.has_starport_unit
+
+    def world_has_zerg_melee_unit(self):
+        if self.world is None:
+            return False
+        else:
+            return self.world.has_zerg_melee_unit
+
+    def world_has_zerg_ranged_unit(self):
+        if self.world is None:
+            return False
+        else:
+            return self.world.has_zerg_ranged_unit
+
+    def world_has_zerg_air_unit(self):
+        if self.world is None:
+            return False
+        else:
+            return self.world.has_zerg_air_unit
+
+    def world_has_protoss_ground_unit(self):
+        if self.world is None:
+            return False
+        else:
+            return self.world.has_protoss_ground_unit
+
+    def world_has_protoss_air_unit(self):
+        if self.world is None:
+            return False
+        else:
+            return self.world.has_protoss_air_unit
+
+    def weapon_armor_upgrade_count(self, upgrade_item: str, state: CollectionState) -> int:
+        assert upgrade_item in upgrade_bundle_inverted_lookup.keys()
+        count: int = 0
+        if self.generic_upgrade_missions > 0:
+            if not self.is_item_placement(state):
+                # Item pool filtering, W/A upgrades aren't items
+                return WEAPON_ARMOR_UPGRADE_MAX_LEVEL
+            else:
+                count += floor(
+                    (100 / self.generic_upgrade_missions)
+                    * (state.count_group("Missions", self.player) / self.total_mission_count())
+                )
+        count += state.count(upgrade_item, self.player)
+        count += state.count_from_list(upgrade_bundle_inverted_lookup[upgrade_item], self.player)
+        if upgrade_item == item_names.PROGRESSIVE_PROTOSS_SHIELDS:
+            count += max(
+                state.count(item_names.PROGRESSIVE_PROTOSS_GROUND_UPGRADE, self.player),
+                state.count(item_names.PROGRESSIVE_PROTOSS_AIR_UPGRADE, self.player)
+            )
+        if (
+                upgrade_item in item_groups.protoss_generic_upgrades
+                and state.has(item_names.QUATRO, self.player)
+        ):
+            count += 1
+        return count
+
+    def terran_army_weapon_armor_upgrade_count(self, state: CollectionState) -> int:
```

Comment:
recommend renaming this and the similar zerg/protoss functions to "X_min_weapon_armour_upgrade_level", as the current name almost implies a total upgrade count rather than a minimum